### DISCUSSION TITLE
How to return the error of `FromForm` to the client?

### DISCUSSION THREAD

**Discussion by duskmoon314** [rwf2/Rocket#2553] [Questions]
TLDR: I want to return the validation result of `FromForm` to the client instead of a 404 message. How can I achieve this?

## Details

I'm trying to build a handler at `get("/path?<format>)`. It can receive several formats and deny other inputs. Thus, I define a struct like this:

```rust
#[derive(FromForm, Debug)]
struct Format<'r>(
    #[field(validate = one_of(...))]
    &'r str,
)
```

This works well, except it prints `expected one of ...` to stdout but returns the message to the client. How should I fix this problem?

**Comment by SergioBenitez** [comment]
I don't totally understand the question, especially this part:

> but returns the message to the client

But maybe what you're looking for is the [`Contextual`](https://api.rocket.rs/v0.5-rc/rocket/form/struct.Contextual.html) form guard. The [`forms` example showcases how to use it](https://github.com/SergioBenitez/Rocket/blob/master/examples/forms/src/main.rs#L73-L87).

Alternatively, if you just want all of the errors, ask for a `Result` with an error types of [`form::Errors`](https://api.rocket.rs/v0.5-rc/rocket/form/struct.Errors.html):

```rust
#[post("/", data = "<form>")]
fn f(data: Result<Form<T>, form::Errors<'_>>) { .. }
```

**Reply by duskmoon314** [reply]
Sorry for not giving a more detailed description.

My question is that the failure of `FromForm` when parsing the query string is printed to the stdout, and the client only gets a `404: NotFound` message.

```rust
#[derive(FromForm, Debug)]
struct Format<'r>(
    #[field(validate = one_of(...))]
    &'r str,
);

// The handler
#[get("/v4?<fmt>")]
fn uuid_v4(fmt: Format) -> Result<String, String> {...}
```

```shell
# stdout
>> Query string failed to match route declaration.
>> expected one of `"hyphenated"`, `"simple"`, `"urn"`, `"braced"`
...
```

```shell
# clients
❯ curl -G http://127.0.0.1:8080/uuid/v4\?fmt\=ur
<!DOCTYPE html>
<html lang="en">
<head>
    <meta charset="utf-8">
    <title>404 Not Found</title>
</head>
<body align="center">
    <div role="main" align="center">
        <h1>404: Not Found</h1>
        <p>The requested resource could not be found.</p>
        <hr />
    </div>
    <div role="contentinfo" align="center">
        <small>Rocket</small>
    </div>
</body>
</html>
```

I want to send the error message to the client instead of printing it to stdout. I will check your suggestions and give you feedback as soon as possible. 😃
